skiko/src/iosMain/kotlin/org/jetbrains/skiko/redrawer/MetalRedrawer.ios.kt

```diff
@@ -116,86 +87,64 @@ internal class MetalRedrawer(
+            val (width, height) = metalLayer.drawableSize.useContents {
+                width.roundToInt() to height.roundToInt()
             }
-        }
-    }
-}
 
-internal class MetalLayer : CAMetalLayer {
-    private lateinit var skiaLayer: SkiaLayer
-    private lateinit var contextHandler: ContextHandler
-
-    @OverrideInit
-    constructor() : super()
-
-    @OverrideInit
-    constructor(layer: Any) : super(layer)
-
-    fun init(
-        skiaLayer: SkiaLayer,
-        contextHandler: ContextHandler,
-        theDevice: MTLDeviceProtocol
-    ) {
-        this.skiaLayer = skiaLayer
-        this.contextHandler = contextHandler
-        this.setNeedsDisplayOnBoundsChange(true)
-        this.removeAllAnimations()
-        // TODO: looks like a bug in K/N interop.
-        this.device = theDevice as objcnames.protocols.MTLDeviceProtocol?
-        this.pixelFormat = MTLPixelFormatBGRA8Unorm
-        this.contentsGravity = kCAGravityTopLeft
-        doubleArrayOf(0.0, 0.0, 0.0, 0.0).usePinned {
-            this.backgroundColor =
-                CGColorCreate(CGColorSpaceCreateDeviceRGB(), it.addressOf(0))
-        }
-        this.framebufferOnly = false
-        this.opaque = false // For UIKit interop through a "Hole"
-        skiaLayer.view?.let {
-            this.frame = it.frame
-            it.layer.addSublayer(this)
-        }
-    }
+            if (width <= 0 || height <= 0) {
+                return@autoreleasepool
+            }
 
-    fun dispose() {
-        this.removeFromSuperlayer()
-        // TODO: anything else to dispose the layer?
-    }
+            dispatch_semaphore_wait(inflightSemaphore, DISPATCH_TIME_FOREVER)
+
+            val metalDrawable = metalLayer.nextDrawable()
+
+            if (metalDrawable == null) {
+                dispatch_semaphore_signal(inflightSemaphore)
+                return@autoreleasepool
+            }
+
+            val renderTarget = BackendRenderTarget.makeMetal(width, height, metalDrawable.texture.objcPtr())
+
+            val surface = Surface.makeFromBackendRenderTarget(
+                context,
+                renderTarget,
+                SurfaceOrigin.TOP_LEFT,
+                SurfaceColorFormat.BGRA_8888,
+                ColorSpace.sRGB,
+                SurfaceProps(pixelGeometry = PixelGeometry.UNKNOWN)
+            )
+
+            if (surface == null) {
+                renderTarget.close()
+                // TODO: manually release metalDrawable when K/N API arrives
+                dispatch_semaphore_signal(inflightSemaphore)
+                return@autoreleasepool
+            }
+
+            surface.canvas.clear(Color.WHITE)
+            drawIntoSurfaceCallback(surface)
+            surface.flushAndSubmit()
 
-    override fun drawInContext(ctx: CGContextRef?) {
-        contextHandler.draw()
-        super.drawInContext(ctx)
+            val commandBuffer = queue.commandBuffer()!!
+            commandBuffer.label = "Present"
```

Comment:
Why `Present`? Can we use something more meaningful? It can be useful for debugging